Clause checks for combined constructs

DavidTruby opened the pull request:
This includes a refactor of the existing combined construct checks
that were present, as well as adding the remaining combined constructs
that had not been implemented yet.

DavidTruby committed:
Added basic clause checks for combined constructs.

This includes a refactor of the existing combined construct checks
that were present, as well as adding the remaining combined constructs
that had not been implemented yet.

ichoyjx commented on test/semantics/omp-clause-validity01.f90:
```diff
@@ -445,7 +445,6 @@
   !$omp end taskloop simd
 
-  !ERROR: REDUCTION clause is not allowed on the TASKLOOP SIMD directive
   !$omp taskloop simd reduction(+:a)
```
No reduction clause can be specified. (4.5 Spec P92)

DavidTruby replied:
Interesting. The reduction clause actually _can_ be specified in 5.0, but not in 4.5. Which is our "source of truth" here? I assume 4.5, in which case I'll add this limitation back in.

ichoyjx commented on lib/semantics/check-omp-structure.cpp:
```diff
@@ -409,6 +556,15 @@ void OmpStructureChecker::Enter(const parser::OpenMPBlockConstruct &x) {
+  case parser::OmpBlockDirective::Directive::TargetParallel: {
+    PushContext(beginDir.source, OmpDirective::TARGET_PARALLEL);
+    SetContextAllowed(
+        (targetAllowedClauses | parallelAllowedClauses) - OmpClause::COPYIN);
```
Could you add a small test for `copyin` clause? Thx.

DavidTruby committed:
Fixed tests for combined constructs.

DavidTruby commented:
@ichoyjx Sorry for letting this rot for so long! I've added some more tests including the one you requested, and I've added back the restriction on reduction clauses on taskloop simd (and the relevant test).
